java/server/src/co/mitro/core/servlets/GetMyPrivateKey.java

```diff
@@ -221,13 +227,43 @@ private void sendEmailAndThrow(GetMyPrivateKeyRequest in, MitroRequestContext co
       String token = makeLoginTokenString(identity, in.extensionId, in.deviceId);
       String tokenSignature = TwoFactorSigningService.signToken(token);
 
-      if (!context.manager.isReadOnly()) {
-        DBEmailQueue email = DBEmailQueue.makeNewDeviceVerification(identity.getName(), token, tokenSignature, context.platform, context.requestServerUrl);
-        context.manager.emailDao.create(email);
-        context.manager.commitTransaction();
+      GenericRawResults<String[]> queuedEmails = context.manager.emailDao.queryRaw(
+              "SELECT type_string, arg_string FROM ("
+              + "(SELECT type_string, arg_string FROM email_queue)"
+              + " UNION "
+              + "(SELECT type_string, arg_string FROM email_queue_sent) "
+              + ") WHERE type_string=? AND arg_string LIKE ?",
+              DBEmailQueue.Type.LOGIN_ON_NEW_DEVICE.getValue(),
+              "%" + identity.getName() + "%"
```

Comment:
Origin should come from the database which in turns comes signing up.
The email_queue table being denormalised, it's quite painful to query. What I could do is assert identity.getName() minus the "%" or "_" characters is over a certain length before doing the query?

People trying to sign in a few times in a row is a common pattern I've seen in the logs so I'd like to make sure we don't flood their inboxes and trigger spam filters.
A new token valid for 12 hours is generated every time. With this PR, users are able to resend 15 minutes after the previous try. We could lower it to 5 minutes if you think it's better.